### ISSUE TITLE
name: Bug Report about: "Numeric (Float64) columns are silently returned as string dtype in synthetic output" labels: bug, severity/major, dtype-regression

### ISSUE THREAD

**Issue by hanifisenturk** [dataxid/dataxid-python#214]
Summary
On valid training input, dataxid.synthesize() and model.generate() occasionally return a DataFrame where a numeric column has been converted to string dtype, even though the training data for that column was clearly numeric (Float64 / Int64). The function does not raise an error, and the output shape and column names are correct, so the defect is a silent correctness bug. Downstream code that does .mean(), .std(), .dt, or any numeric operation on the affected column will either crash or — worse — produce lexicographic results instead of numeric ones.

DataXID SDK: pip install dataxid (beta window 2026-04)
Python: 3.11
pandas: 2.2.x
Reproduced on: Google Colab (Ubuntu) and local macOS 14 (+2 multi-env bonus)

Steps to Reproduce
Two independent reproductions from my test run, both triggering the same defect:
Reproduction A — Iris, via Model.create → generate
pythonimport dataxid
from sklearn.datasets import load_iris

dataxid.api_key = "dx_test_..."

df = load_iris(as_frame=True).frame.head(100)
print("Input dtypes:"); print(df.dtypes)
# sepal length (cm)    float64
# sepal width (cm)     float64
# petal length (cm)    float64
# petal width (cm)     float64
# target                 int64

cfg = dataxid.ModelConfig(max_epochs=50, batch_size=32, model_size="large")
model = dataxid.Model.create(data=df, config=cfg)
synthetic = model.generate(n_samples=1000)

print("Output dtypes:"); print(synthetic.dtypes)
# sepal length (cm)    Float64
# sepal width (cm)     Float64
# petal length (cm)    Float64
# petal width (cm)     string   <-- SHOULD BE Float64
# target                 Int64

synthetic["petal width (cm)"].mean()
# raises: cannot perform mean on object/string dtype
Reproduction B — mixed frame with a constant column
pythondf = pd.DataFrame({
    "constant_column":    [1, 1, 1, 1, 1],
    "numeric_column":     [1.0, 2.0, 3.0, 4.0, 5.0],
    "categorical_column": ["a", "b", "c", "d", "e"],
})

synthetic = dataxid.synthesize(
    data=df,
    n_samples=100,
    config=dataxid.ModelConfig(max_epochs=100, batch_size=10, model_size="medium"),
)
print(synthetic.dtypes)
# constant_column        Int64
# numeric_column        string   <-- SHOULD BE Float64
# categorical_column    string
Note that both happy-path runs (e.g., Titanic with 1000 rows) preserve dtypes correctly — the defect is not 100% deterministic, it is correlated with specific dataset characteristics.
Expected Behavior
The synthetic DataFrame's dtypes must match the training DataFrame's dtypes column-for-column. A Float64 column in the input must come back as Float64 in the output. This is implicit in the challenge manual's claim that the SDK "produces artificial data with the same distribution" — a string cannot share a distribution with a float.
Actual Behavior
One numeric column per run is converted to string. Across my twelve-scenario test run, this appeared on at least two independent scenarios (Functional Specialist iter 3 on Iris, and Edge-Case Warrior iter 3 on the constant-column mix). Other runs preserved dtypes correctly, which rules out a global "always strings" bug and points at a dataset-dependent code path.
Root Cause (hypothesis)
The most likely cause is that the SDK's generation step produces each column independently and, for some columns, the generator returns categorical samples that are then stored as strings without being coerced back to the original numeric dtype. A likely trigger is any column whose distribution collapses to a small number of discrete values at generation time (e.g., petal width (cm) in iris has only ~20 unique values, so it may be routed to the categorical encoder). A correct implementation should record the input dtypes at fit time and apply astype() — with pd.to_numeric(..., errors="coerce") as a fallback — to the output before returning.
Suggested Fix
python# In dataxid/_synthesize.py, at the end of synthesize() / generate():
original_dtypes = data.dtypes.to_dict()
for col, dt in original_dtypes.items():
    if col in result.columns and str(result[col].dtype) != str(dt):
        if pd.api.types.is_numeric_dtype(dt):
            result[col] = pd.to_numeric(result[col], errors="coerce").astype(dt)
        else:
            result[col] = result[col].astype(dt, errors="ignore")
return result
Plus a regression test that asserts synthetic.dtypes.equals(df.dtypes) across all three sample datasets (Adult, Iris, Titanic).

**Comment by BurakUlver** [comment]
This is expected with auto encoding: low-cardinality numeric columns can be treated as categorical. Pass encoding_types to force a numeric encoding when you need Float64 out. Closing as working as designed.
